R/read_portfolio_csv.R

```diff
@@ -135,8 +52,110 @@ read_portfolio_csv <- function(filepaths, combine = TRUE) {
   }
 
   if (combine) {
-    return(dplyr::bind_rows(portfolio_dfs[!is.na(portfolio_dfs)], .id = "filepath"))
+    return(
+      dplyr::bind_rows(portfolio_dfs[!is.na(portfolio_dfs)], .id = "filepath")
+    )
   }
 
   portfolio_dfs
 }
+
+read_single_portfolio_csv <- function(filepath) {
+
+  if (length(filepath) != 1L) {
+    stop("`filepath` must be a single string")
+  }
+  if (typeof(filepath) != "character") {
+    stop("`filepath` must be a string")
+  }
+  if (!file.exists(filepath)) {
+    stop(paste("file does not exist:", filepath))
+  }
+
+  if (!is_text_file(filepath)) {
+    return(list(NA))
+  }
+
+  if (!has_consistent_fields_per_line(filepath)) {
+    return(list(NA))
+  }
+
+  encoding <- guess_file_encoding(filepath)
+  delimiter <- guess_delimiter(filepath)
+  decimal_mark <- guess_decimal_mark(filepath)
+  grouping_mark <- guess_grouping_mark(filepath)
+  if (any(is.na(c(encoding, delimiter, decimal_mark, grouping_mark)))) {
+    return(list(NA))
+  }
```

Comment:
Given the context of this function, I think maybe these failure points should cause a stop/error versus returning an `NA`

Reply:
like, if any of these tests fail, it cannot proceed, and since we're only reading one file, the `NA` to fill a list so that the output (list of read in data) is the same size as the input (vector of filepaths) is not helpful

Reply:
though, now that I've said that I'm realizing that would probably break `read_portfolio_csvs()`, so maybe don't do that

Reply:
Yeah, my goal with this PR is to keep the logic as-is, and then we can sort out stop conditions later. In general, I think I'd rather avoid a `stop()` in the body (other than the checks on singular filepath) since that would force a `tryCatch()` elsewhere.

my general thought on expanding this is to have something like
```R
read_single_portfolio_csv <- function(
  filepath,
  strict = TRUE # current behavior
  messages = !strict
) {
  #Blah Blah Blah producing informative messages
  if (messages) {
    return(list(portfolio = portfolio_df, messages = messages))
  } else {
    return(list(portfolio_df))
  }
}
```

which would allow us to raise informative info around _what_ is wrong with a particular file.